diff --git a/src/serverenvironment.cpp b/src/serverenvironment.cpp
index 3d9ba132b..99e815bd4 100644
--- a/src/serverenvironment.cpp
+++ b/src/serverenvironment.cpp
@@ -1359,7 +1359,7 @@ void ServerEnvironment::step(float dtime)
 				continue;

 			// Set current time as timestamp (and let it set ChangedFlag)
-			block->setTimestamp(m_game_time);
+			//block->setTimestamp(m_game_time);
 		}

 		/*
@@ -1398,9 +1398,11 @@ void ServerEnvironment::step(float dtime)
 			block->setTimestampNoChangedFlag(m_game_time);
 			// If time has changed much from the one on disk,
 			// set block to be saved when it is unloaded
+			/*
 			if(block->getTimestamp() > block->getDiskTimestamp() + 60)
 				block->raiseModified(MOD_STATE_WRITE_AT_UNLOAD,
 					MOD_REASON_BLOCK_EXPIRED);
+			*/

 			// Run node timers
 			std::vector<NodeTimer> elapsed_timers = block->m_node_timers.step(dtime);
