diff --git a/src/script/lua_api/l_server.cpp b/src/script/lua_api/l_server.cpp
index 00e849cd..671ece0e 100644
--- a/src/script/lua_api/l_server.cpp
+++ b/src/script/lua_api/l_server.cpp
@@ -26,6 +26,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "environment.h"
 #include "remoteplayer.h"
 #include "log.h"
+#include "profiler.h"
 #include <algorithm>
 
 // request_shutdown()
@@ -395,6 +396,16 @@ int ModApiServer::l_get_modpath(lua_State *L)
 	return 1;
 }
 
+// get_profiler_value(profiler_name)
+int ModApiServer::l_get_profiler_value(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	std::string profiler_name = luaL_checkstring(L, 1);
+	float value = g_profiler->getValue(profiler_name);
+	lua_pushnumber(L, value);
+	return 1;
+}
+
 // get_modnames()
 // the returned list is sorted alphabetically for you
 int ModApiServer::l_get_modnames(lua_State *L)
@@ -521,6 +532,7 @@ void ModApiServer::Initialize(lua_State *L, int top)
 
 	API_FCT(get_current_modname);
 	API_FCT(get_modpath);
+	API_FCT(get_profiler_value);
 	API_FCT(get_modnames);
 
 	API_FCT(print);
diff --git a/src/script/lua_api/l_server.h b/src/script/lua_api/l_server.h
index 3aa1785a..c04d48ee 100644
--- a/src/script/lua_api/l_server.h
+++ b/src/script/lua_api/l_server.h
@@ -45,6 +45,9 @@ class ModApiServer : public ModApiBase
 	// get_modpath(modname)
 	static int l_get_modpath(lua_State *L);
 
+	// get_profiler_value(profiler_name)
+	static int l_get_profiler_value(lua_State *L);
+
 	// get_modnames()
 	// the returned list is sorted alphabetically for you
 	static int l_get_modnames(lua_State *L);
