diff --git a/src/database/database-sqlite3.cpp b/src/database/database-sqlite3.cpp
index 1bacdfe6..5bdd5b22 100644
--- a/src/database/database-sqlite3.cpp
+++ b/src/database/database-sqlite3.cpp
@@ -659,7 +659,7 @@ void AuthDatabaseSQLite3::initStatements()
 {
 	PREPARE_STATEMENT(read, "SELECT id, name, password, last_login FROM auth WHERE name = ?");
 	PREPARE_STATEMENT(write, "UPDATE auth set name = ?, password = ?, last_login = ? WHERE id = ?");
-	PREPARE_STATEMENT(create, "INSERT INTO auth (name, password, last_login) VALUES (?, ?, ?)");
+	PREPARE_STATEMENT(create, "INSERT OR REPLACE INTO auth (name, password, last_login) VALUES (?, ?, ?)");
 	PREPARE_STATEMENT(delete, "DELETE FROM auth WHERE name = ?");
 
 	PREPARE_STATEMENT(list_names, "SELECT name FROM auth ORDER BY name DESC");
