diff --git a/src/database/CMakeLists.txt b/src/database/CMakeLists.txt
index e9d157c29..839c9152f 100644
--- a/src/database/CMakeLists.txt
+++ b/src/database/CMakeLists.txt
@@ -4,6 +4,8 @@ set(database_SRCS
 	${CMAKE_CURRENT_SOURCE_DIR}/database-files.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/database-leveldb.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/database-postgresql.cpp
+	${CMAKE_CURRENT_SOURCE_DIR}/postgres/mapsavequeue.cpp
+	${CMAKE_CURRENT_SOURCE_DIR}/postgres/playersavequeue.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/database-redis.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/database-sqlite3.cpp
 	PARENT_SCOPE
diff --git a/src/database/database-postgresql.cpp b/src/database/database-postgresql.cpp
index ca750b466..c65e14bd5 100644
--- a/src/database/database-postgresql.cpp
+++ b/src/database/database-postgresql.cpp
@@ -176,8 +176,12 @@ MapDatabasePostgreSQL::MapDatabasePostgreSQL(const std::string &connect_string):
 	MapDatabase()
 {
 	connectToDatabase();
+  map_save_queue = new MapSaveQueue(connect_string);
 }
 
+MapDatabasePostgreSQL::~MapDatabasePostgreSQL() {
+  delete map_save_queue;
+}
 
 void MapDatabasePostgreSQL::createDatabase()
 {
@@ -238,6 +242,8 @@ bool MapDatabasePostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
 		return false;
 	}
 
+  map_save_queue->enqueue(pos, data);
+  /*
 	pingDatabase();
 
 	s32 x, y, z;
@@ -257,6 +263,7 @@ bool MapDatabasePostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
 	} else {
 		execPrepared("write_block", ARRLEN(args), args, argLen, argFmt);
 	}
+  */
 	return true;
 }
 
@@ -325,6 +332,11 @@ PlayerDatabasePostgreSQL::PlayerDatabasePostgreSQL(const std::string &connect_st
 	PlayerDatabase()
 {
 	connectToDatabase();
+  player_save_queue = new PlayerSaveQueue(connect_string);
+}
+
+PlayerDatabasePostgreSQL::~PlayerDatabasePostgreSQL(){
+  delete player_save_queue;
 }
 
 
@@ -462,6 +474,8 @@ void PlayerDatabasePostgreSQL::savePlayer(RemotePlayer *player)
 	if (!sao)
 		return;
 
+  player_save_queue->enqueue(player);
+  /*
 	pingDatabase();
 
 	v3f pos = sao->getBasePosition();
@@ -539,7 +553,7 @@ void PlayerDatabasePostgreSQL::savePlayer(RemotePlayer *player)
 		execPrepared("save_player_metadata", 3, meta_values);
 	}
 	endSave();
-
+  */
 	player->onSuccessfulSave();
 }
 
diff --git a/src/database/database-postgresql.h b/src/database/database-postgresql.h
index 340f0a7b8..c3fd24e07 100644
--- a/src/database/database-postgresql.h
+++ b/src/database/database-postgresql.h
@@ -23,6 +23,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include <libpq-fe.h>
 #include "database.h"
 #include "util/basic_macros.h"
+#include "postgres/mapsavequeue.h"
+#include "postgres/playersavequeue.h"
+#include "postgres/pgutil.h"
 
 class Settings;
 
@@ -112,7 +115,7 @@ class MapDatabasePostgreSQL : private Database_PostgreSQL, public MapDatabase
 {
 public:
 	MapDatabasePostgreSQL(const std::string &connect_string);
-	virtual ~MapDatabasePostgreSQL() = default;
+	virtual ~MapDatabasePostgreSQL();
 
 	virtual void pingDatabase() { Database_PostgreSQL::pingDatabase(); }
 
@@ -124,6 +127,8 @@ class MapDatabasePostgreSQL : private Database_PostgreSQL, public MapDatabase
 	void beginSave() { Database_PostgreSQL::beginSave(); }
 	void endSave() { Database_PostgreSQL::endSave(); }
 
+	MapSaveQueue *map_save_queue;
+
 protected:
 	virtual void createDatabase();
 	virtual void initStatements();
@@ -133,7 +138,7 @@ class PlayerDatabasePostgreSQL : private Database_PostgreSQL, public PlayerDatab
 {
 public:
 	PlayerDatabasePostgreSQL(const std::string &connect_string);
-	virtual ~PlayerDatabasePostgreSQL() = default;
+	virtual ~PlayerDatabasePostgreSQL();
 
 	virtual void pingDatabase() { Database_PostgreSQL::pingDatabase(); }
 
@@ -142,6 +147,8 @@ class PlayerDatabasePostgreSQL : private Database_PostgreSQL, public PlayerDatab
 	bool removePlayer(const std::string &name);
 	void listPlayers(std::vector<std::string> &res);
 
+	PlayerSaveQueue *player_save_queue;
+
 protected:
 	virtual void createDatabase();
 	virtual void initStatements();
diff --git a/src/database/postgres/mapsavequeue.cpp b/src/database/postgres/mapsavequeue.cpp
new file mode 100644
index 000000000..e5fa98b3e
--- /dev/null
+++ b/src/database/postgres/mapsavequeue.cpp
@@ -0,0 +1,160 @@
+
+#include "config.h"
+
+#ifdef USE_POSTGRESQL
+
+#include "mapsavequeue.h"
+
+#include "log.h"
+#include "threading/mutex_auto_lock.h"
+#include "exceptions.h"
+
+#include <thread>
+#include <chrono>
+#include <list>
+#include "pgutil.h"
+
+MapSaveQueue::MapSaveQueue(const std::string &connect_string) :
+	Thread("map_save_queue") {
+
+	m_conn = PQconnectdb(connect_string.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgversion = PQserverVersion(m_conn);
+
+	if (m_pgversion < 90500) {
+		PGUtil::prepareStatement(m_conn, "write_block_insert",
+			"INSERT INTO blocks (posX, posY, posZ, data) SELECT "
+				"$1::int4, $2::int4, $3::int4, $4::bytea "
+				"WHERE NOT EXISTS (SELECT true FROM blocks "
+				"WHERE posX = $1::int4 AND posY = $2::int4 AND "
+				"posZ = $3::int4)");
+
+		PGUtil::prepareStatement(m_conn, "write_block_update",
+			"UPDATE blocks SET data = $4::bytea "
+				"WHERE posX = $1::int4 AND posY = $2::int4 AND "
+				"posZ = $3::int4");
+	} else {
+		PGUtil::prepareStatement(m_conn, "write_block",
+			"INSERT INTO blocks (posX, posY, posZ, data) VALUES "
+				"($1::int4, $2::int4, $3::int4, $4::bytea) "
+				"ON CONFLICT ON CONSTRAINT blocks_pkey DO "
+				"UPDATE SET data = $4::bytea");
+	}
+
+}
+
+MapSaveQueue::~MapSaveQueue(){}
+
+void *MapSaveQueue::run(){
+	std::vector<QueuedItem*> save_items;
+
+	while (!stopRequested()){
+		std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+		// move items over here
+		MutexAutoLock lock(m_mutex);
+		if (!queue.empty()){
+			save_items.swap(queue);
+		}
+
+		this->save(&save_items);
+		save_items.clear();
+	}
+
+
+	if (!queue.empty()){
+		//flush at exit
+		MutexAutoLock lock(m_mutex);
+		this->save(&queue);
+	}
+
+	return nullptr;
+}
+
+void MapSaveQueue::saveBlock(QueuedItem *item){
+	// Verify if we don't overflow the platform integer with the mapblock size
+	if (item->data.size() > INT_MAX) {
+		errorstream << "Database_PostgreSQL::saveBlock: Data truncation! "
+		<< "data.size() over 0xFFFFFFFF (== " << item->data.size()
+		<< ")" << std::endl;
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: data truncation @ " +
+			std::to_string(item->pos.X) + "/" +
+			std::to_string(item->pos.Y) + "/" +
+			std::to_string(item->pos.Z)
+			)
+		);
+	}
+
+	if (PQstatus(m_conn) != CONNECTION_OK){
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn)
+		);
+	}
+
+	s32 x, y, z;
+	x = htonl(item->pos.X);
+	y = htonl(item->pos.Y);
+	z = htonl(item->pos.Z);
+
+	const void *args[] = { &x, &y, &z, item->data.c_str() };
+	const int argLen[] = {
+		sizeof(x), sizeof(y), sizeof(z), (int)(item->data.size())
+	};
+	const int argFmt[] = { 1, 1, 1, 1 };
+
+	if (m_pgversion < 90500) {
+		PGUtil::execPrepared(m_conn, "write_block_update", ARRLEN(args), args, argLen, argFmt);
+		PGUtil::execPrepared(m_conn, "write_block_insert", ARRLEN(args), args, argLen, argFmt);
+	} else {
+		PGUtil::execPrepared(m_conn, "write_block", ARRLEN(args), args, argLen, argFmt);
+	}
+
+}
+
+void MapSaveQueue::save(std::vector<QueuedItem*> *items){
+
+	try {
+		PGUtil::checkResults(PQexec(m_conn, "BEGIN;"));
+
+		for (QueuedItem* item: *items){
+			saveBlock(item);
+			delete item;
+		}
+
+		PGUtil::checkResults(PQexec(m_conn, "COMMIT;"));
+
+	} catch (std::exception &e) {
+		MutexAutoLock lock(m_exception_mutex);
+		m_async_exception = std::current_exception();
+		this->stop();
+
+	}
+}
+
+
+void MapSaveQueue::enqueue(const v3s16 &pos, const std::string &data){
+	{
+		MutexAutoLock lock(m_exception_mutex);
+		if (m_async_exception) {
+			std::rethrow_exception(m_async_exception);
+		}
+	}
+
+	MutexAutoLock lock(m_mutex);
+
+	QueuedItem *item = new QueuedItem();
+	item->pos = pos;
+	item->data = data;
+
+	queue.push_back(item);
+}
+
+#endif //USE_POSTGRESQL
diff --git a/src/database/postgres/mapsavequeue.h b/src/database/postgres/mapsavequeue.h
new file mode 100644
index 000000000..93b71093c
--- /dev/null
+++ b/src/database/postgres/mapsavequeue.h
@@ -0,0 +1,67 @@
+/*
+Minetest
+Copyright (C) 2019 BuckarooBanzai/naturefreshmilk, Thomas Rudin <thomas@rudin.io>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include <string>
+#include "database/database.h"
+#include "threading/thread.h"
+#include "irr_v3d.h"
+#include <libpq-fe.h>
+
+// for htonl()
+#ifdef _WIN32
+        // Without this some of the network functions are not found on mingw
+        #ifndef _WIN32_WINNT
+                #define _WIN32_WINNT 0x0501
+        #endif
+        #include <windows.h>
+        #include <winsock2.h>
+#else
+	#include <netinet/in.h>
+#endif
+
+
+struct QueuedItem {
+	v3s16 pos;
+	std::string data;
+};
+
+class MapSaveQueue : public Thread {
+public:
+	MapSaveQueue(const std::string &connect_string);
+	~MapSaveQueue();
+	void enqueue(const v3s16 &pos, const std::string &data);
+
+protected:
+	void *run();
+
+private:
+	PGconn *m_conn = nullptr;
+	std::exception_ptr m_async_exception;
+	int m_pgversion;
+	void save(std::vector<QueuedItem*> *items);
+	void saveBlock(QueuedItem *item);
+
+	std::mutex m_mutex;
+	std::mutex m_exception_mutex;
+
+	std::vector<QueuedItem*> queue;
+
+};
diff --git a/src/database/postgres/pgutil.h b/src/database/postgres/pgutil.h
new file mode 100644
index 000000000..8010e873c
--- /dev/null
+++ b/src/database/postgres/pgutil.h
@@ -0,0 +1,94 @@
+/*
+Minetest
+Copyright (C) 2019 BuckarooBanzai/naturefreshmilk, Thomas Rudin <thomas@rudin.io>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+#pragma once
+
+#include <libpq-fe.h>
+#include "irr_v3d.h"
+#include "exceptions.h"
+
+namespace PGUtil {
+
+	inline PGresult *checkResults(PGresult *result, bool clear = true)
+	{
+		ExecStatusType statusType = PQresultStatus(result);
+
+		switch (statusType) {
+		case PGRES_COMMAND_OK:
+		case PGRES_TUPLES_OK:
+			break;
+		case PGRES_FATAL_ERROR:
+		default:
+			throw DatabaseException(
+				std::string("PostgreSQL database error: ") +
+				PQresultErrorMessage(result));
+		}
+
+		if (clear)
+			PQclear(result);
+
+		return result;
+	}
+
+	inline void prepareStatement(PGconn *m_conn, const std::string &name, const std::string &sql)
+	{
+		checkResults(PQprepare(m_conn, name.c_str(), sql.c_str(), 0, NULL));
+	}
+
+	inline PGresult *execPrepared(PGconn *m_conn, const char *stmtName, const int paramsNumber,
+		const void **params,
+		const int *paramsLengths = NULL, const int *paramsFormats = NULL,
+		bool clear = true, bool nobinary = true) {
+
+		return checkResults(PQexecPrepared(m_conn, stmtName, paramsNumber,
+			(const char* const*) params, paramsLengths, paramsFormats,
+			nobinary ? 1 : 0), clear);
+	}
+
+	inline PGresult *execPrepared(PGconn *m_conn, const char *stmtName, const int paramsNumber,
+		const char **params, bool clear = true, bool nobinary = true) {
+		return execPrepared(m_conn, stmtName, paramsNumber,
+			(const void **)params, NULL, NULL, clear, nobinary);
+	}
+
+	// Conversion helpers
+	inline int pg_to_int(PGresult *res, int row, int col)
+	{
+		return atoi(PQgetvalue(res, row, col));
+	}
+
+	inline u32 pg_to_uint(PGresult *res, int row, int col)
+	{
+		return (u32) atoi(PQgetvalue(res, row, col));
+	}
+
+	inline float pg_to_float(PGresult *res, int row, int col)
+	{
+		return (float) atof(PQgetvalue(res, row, col));
+	}
+
+	inline v3s16 pg_to_v3s16(PGresult *res, int row, int col)
+	{
+		return v3s16(
+			pg_to_int(res, row, col),
+			pg_to_int(res, row, col + 1),
+			pg_to_int(res, row, col + 2)
+		);
+	}
+
+}
\ No newline at end of file
diff --git a/src/database/postgres/playersavequeue.cpp b/src/database/postgres/playersavequeue.cpp
new file mode 100644
index 000000000..b629acceb
--- /dev/null
+++ b/src/database/postgres/playersavequeue.cpp
@@ -0,0 +1,324 @@
+
+#include "config.h"
+
+#if USE_POSTGRESQL
+
+#include "playersavequeue.h"
+
+#include "log.h"
+#include "threading/mutex_auto_lock.h"
+#include "exceptions.h"
+#include "pgutil.h"
+#include "server/player_sao.h"
+
+#include <thread>
+#include <chrono>
+#include <list>
+
+
+/*
+Serialized player data
+*/
+
+SerializedPlayer::SerializedPlayer(RemotePlayer *player){
+	PlayerSAO* sao = player->getPlayerSAO();
+	v3f pos = sao->getBasePosition();
+
+	name = player->getName();
+	pitch = ftos(sao->getLookPitch());
+	yaw = ftos(sao->getRotation().Y);
+	posx = ftos(pos.X);
+	posy = ftos(pos.Y);
+	posz = ftos(pos.Z);
+	hp = itos(sao->getHP());
+	breath = itos(sao->getBreath());
+}
+
+void SerializedPlayer::persist(PGconn *m_conn, int pgversion){
+	const char *values[] = {
+		name.c_str(),
+		pitch.c_str(),
+		yaw.c_str(),
+		posx.c_str(), posy.c_str(), posz.c_str(),
+		hp.c_str(),
+		breath.c_str()
+	};
+
+	if (pgversion < 90500) {
+
+		const char *exists_values[] = { name.c_str() };
+		PGresult *exists_results = PGUtil::execPrepared(m_conn, "load_player", 1, exists_values, false);
+
+		bool player_exists = (PQntuples(exists_results) > 0);
+		PQclear(exists_results);
+
+		if (!player_exists)
+			PGUtil::execPrepared(m_conn, "create_player", 8, values, true, false);
+		else
+			PGUtil::execPrepared(m_conn, "update_player", 8, values, true, false);
+	} else {
+		PGUtil::execPrepared(m_conn, "save_player", 8, values, true, false);
+	}
+
+}
+
+std::string SerializedPlayer::getName(){
+	return name;
+}
+
+/*
+Serialized inventory data
+*/
+
+SerializedInventory::SerializedInventory(std::string player, std::string inv_id, const InventoryList* list){
+	this->player = player;
+	this->inv_id = inv_id;
+
+	name = list->getName();
+	width = itos(list->getWidth());
+	lsize = itos(list->getSize());
+}
+
+void SerializedInventory::persist(PGconn *m_conn){
+	const char* inv_values[] = {
+		player.c_str(),
+		inv_id.c_str(),
+		width.c_str(),
+		name.c_str(),
+		lsize.c_str()
+	};
+
+	PGUtil::execPrepared(m_conn, "add_player_inventory", 5, inv_values);
+}
+
+/*
+Serialized inventory item data
+*/
+
+SerializedInventoryItem::SerializedInventoryItem(std::string player, std::string inv_id, std::string slotId, std::string itemStr):
+	player(player), inv_id(inv_id), slotId(slotId), itemStr(itemStr)
+{}
+
+void SerializedInventoryItem::persist(PGconn *m_conn){
+	const char* invitem_values[] = {
+		player.c_str(),
+		inv_id.c_str(),
+		slotId.c_str(),
+		itemStr.c_str()
+	};
+	PGUtil::execPrepared(m_conn, "add_player_inventory_item", 4, invitem_values);
+}
+
+/*
+Serializes player metadata
+*/
+
+SerializedPlayerMetadata::SerializedPlayerMetadata(std::string player, std::string attr, std::string value):
+	player(player), attr(attr), value(value)
+{}
+
+void SerializedPlayerMetadata::persist(PGconn *m_conn){
+	const char *meta_values[] = {
+		player.c_str(),
+		attr.c_str(),
+		value.c_str()
+	};
+	PGUtil::execPrepared(m_conn, "save_player_metadata", 3, meta_values);
+}
+
+/*
+Save queue
+*/
+
+PlayerSaveQueue::PlayerSaveQueue(const std::string &connect_string) :
+	Thread("player_save_queue") {
+
+	m_conn = PQconnectdb(connect_string.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) {
+		throw DatabaseException(std::string(
+			"PostgreSQL database error: ") +
+			PQerrorMessage(m_conn));
+	}
+
+	m_pgversion = PQserverVersion(m_conn);
+
+	if (m_pgversion < 90500) {
+		PGUtil::prepareStatement(m_conn, "create_player",
+			"INSERT INTO player(name, pitch, yaw, posX, posY, posZ, hp, breath) VALUES "
+				"($1, $2, $3, $4, $5, $6, $7::int, $8::int)");
+
+		PGUtil::prepareStatement(m_conn, "update_player",
+			"UPDATE SET pitch = $2, yaw = $3, posX = $4, posY = $5, posZ = $6, hp = $7::int, "
+				"breath = $8::int, modification_date = NOW() WHERE name = $1");
+	} else {
+		PGUtil::prepareStatement(m_conn, "save_player",
+			"INSERT INTO player(name, pitch, yaw, posX, posY, posZ, hp, breath) VALUES "
+				"($1, $2, $3, $4, $5, $6, $7::int, $8::int)"
+				"ON CONFLICT ON CONSTRAINT player_pkey DO UPDATE SET pitch = $2, yaw = $3, "
+				"posX = $4, posY = $5, posZ = $6, hp = $7::int, breath = $8::int, "
+				"modification_date = NOW()");
+	}
+
+	PGUtil::prepareStatement(m_conn, "remove_player_inventories",
+		"DELETE FROM player_inventories WHERE player = $1");
+
+	PGUtil::prepareStatement(m_conn, "remove_player_inventory_items",
+		"DELETE FROM player_inventory_items WHERE player = $1");
+
+	PGUtil::prepareStatement(m_conn, "add_player_inventory",
+		"INSERT INTO player_inventories (player, inv_id, inv_width, inv_name, inv_size) VALUES "
+			"($1, $2::int, $3::int, $4, $5::int)");
+
+	PGUtil::prepareStatement(m_conn, "add_player_inventory_item",
+		"INSERT INTO player_inventory_items (player, inv_id, slot_id, item) VALUES "
+			"($1, $2::int, $3::int, $4)");
+
+	PGUtil::prepareStatement(m_conn, "remove_player_metadata",
+		"DELETE FROM player_metadata WHERE player = $1");
+
+	PGUtil::prepareStatement(m_conn, "save_player_metadata",
+		"INSERT INTO player_metadata (player, attr, value) VALUES ($1, $2, $3)");
+
+}
+
+PlayerSaveQueue::~PlayerSaveQueue(){
+}
+
+
+void PlayerSaveQueue::enqueue(RemotePlayer *player){
+	{
+		MutexAutoLock lock(m_exception_mutex);
+		if (m_async_exception) {
+			std::rethrow_exception(m_async_exception);
+		}
+	}
+
+	MutexAutoLock lock(m_mutex);
+
+	QueuedPlayerData *queued_data = new QueuedPlayerData;
+	queued_data->serialized_player = new SerializedPlayer(player);
+
+	std::string playername = player->getName();
+	PlayerSAO* sao = player->getPlayerSAO();
+
+	std::vector<const InventoryList*> inventory_lists = sao->getInventory()->getLists();
+	for (u16 i = 0; i < inventory_lists.size(); i++) {
+		const InventoryList* list = inventory_lists[i];
+		std::string inv_id = itos(i);
+
+		SerializedInventory* serialized_inv = new SerializedInventory(playername, inv_id, list);
+		queued_data->inventories.push_back(serialized_inv);
+
+
+		for (u32 j = 0; j < list->getSize(); j++) {
+			std::ostringstream os;
+			list->getItem(j).serialize(os);
+			std::string itemStr = os.str();
+			std::string slotId = itos(j);
+
+			SerializedInventoryItem *invItem = new SerializedInventoryItem(
+				playername,
+				inv_id,
+				slotId,
+				itemStr
+			);
+
+			queued_data->inventory_items.push_back(invItem);
+		}
+	}
+
+	const StringMap &attrs = sao->getMeta().getStrings();
+	for (const auto &attr : attrs) {
+
+		SerializedPlayerMetadata *metadata = new SerializedPlayerMetadata(
+			playername,
+			attr.first,
+			attr.second
+		);
+
+		queued_data->metadata.push_back(metadata);
+	}
+
+	queue.push_back(queued_data);
+}
+
+
+void *PlayerSaveQueue::run(){
+	std::vector<QueuedPlayerData*> save_items;
+
+	while (!stopRequested()){
+		std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+		{
+			// move items over here
+			MutexAutoLock lock(m_mutex);
+			if (!queue.empty()){
+				save_items.swap(queue);
+			}
+		}
+
+		this->save(&save_items);
+		save_items.clear();
+	}
+
+
+	if (!queue.empty()){
+		//flush at exit
+		MutexAutoLock lock(m_mutex);
+		this->save(&queue);
+	}
+
+	return nullptr;
+}
+
+void PlayerSaveQueue::save(std::vector<QueuedPlayerData*> *queue){
+	try {
+
+		for (QueuedPlayerData* item: *queue){
+			PGUtil::checkResults(PQexec(m_conn, "BEGIN;"));
+			savePlayer(item);
+			PGUtil::checkResults(PQexec(m_conn, "COMMIT;"));
+			delete item;
+		}
+
+
+	} catch (std::exception &e) {
+		MutexAutoLock lock(m_exception_mutex);
+		m_async_exception = std::current_exception();
+		this->stop();
+
+	}
+}
+
+void PlayerSaveQueue::savePlayer(QueuedPlayerData *item){
+
+	std::string playername = item->serialized_player->getName();
+	const char* rmvalues[] = { playername.c_str() };
+
+	item->serialized_player->persist(m_conn, m_pgversion);
+
+	// Write player inventories
+	PGUtil::execPrepared(m_conn, "remove_player_inventories", 1, rmvalues);
+	PGUtil::execPrepared(m_conn, "remove_player_inventory_items", 1, rmvalues);
+	PGUtil::execPrepared(m_conn, "remove_player_metadata", 1, rmvalues);
+
+	for (SerializedInventory *inv: item->inventories){
+		inv->persist(m_conn);
+		delete inv;
+	}
+
+	for (SerializedInventoryItem *invItem: item->inventory_items){
+		invItem->persist(m_conn);
+		delete invItem;
+	}
+
+	for (SerializedPlayerMetadata *metadata: item->metadata){
+		metadata->persist(m_conn);
+		delete metadata;
+	}
+
+	delete item->serialized_player;
+}
+
+#endif //USE_POSTGRESQL
diff --git a/src/database/postgres/playersavequeue.h b/src/database/postgres/playersavequeue.h
new file mode 100644
index 000000000..3ba71dfa7
--- /dev/null
+++ b/src/database/postgres/playersavequeue.h
@@ -0,0 +1,109 @@
+/*
+Minetest
+Copyright (C) 2019 BuckarooBanzai/naturefreshmilk, Thomas Rudin <thomas@rudin.io>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include <string>
+#include "database/database.h"
+#include "threading/thread.h"
+#include <libpq-fe.h>
+#include "remoteplayer.h"
+
+class SerializedPlayer {
+public:
+	SerializedPlayer(RemotePlayer *player);
+	void persist(PGconn *m_conn, int pgversion);
+	std::string getName();
+private:
+	std::string name;
+	std::string pitch;
+	std::string yaw;
+	std::string posx;
+	std::string posy;
+	std::string posz;
+	std::string hp;
+	std::string breath;
+};
+
+class SerializedInventory {
+public:
+	SerializedInventory(std::string player, std::string inv_id, const InventoryList* list);
+	void persist(PGconn *m_conn);
+private:
+	std::string player;
+	std::string inv_id;
+	std::string width;
+	std::string name;
+	std::string lsize;
+};
+
+class SerializedInventoryItem {
+public:
+	SerializedInventoryItem(std::string player, std::string inv_id, std::string slotId, std::string itemStr);
+	void persist(PGconn *m_conn);
+private:
+	std::string player;
+	std::string inv_id;
+	std::string slotId;
+	std::string itemStr;
+};
+
+class SerializedPlayerMetadata {
+public:
+	SerializedPlayerMetadata(std::string player, std::string attr, std::string value);
+	void persist(PGconn *m_conn);
+private:
+	std::string player;
+	std::string attr;
+	std::string value;
+
+};
+
+struct QueuedPlayerData {
+	SerializedPlayer *serialized_player;
+	std::vector<SerializedInventory*> inventories;
+	std::vector<SerializedInventoryItem*> inventory_items;
+	std::vector<SerializedPlayerMetadata*> metadata;
+};
+
+
+class PlayerSaveQueue : public Thread {
+public:
+	PlayerSaveQueue(const std::string &connect_string);
+	~PlayerSaveQueue();
+	void enqueue(RemotePlayer *player);
+
+protected:
+	void *run();
+
+private:
+	std::exception_ptr m_async_exception;
+
+	int m_pgversion;
+
+	void save(std::vector<QueuedPlayerData*> *queue);
+	void savePlayer(QueuedPlayerData *playerdata);
+
+	PGconn *m_conn = nullptr;
+
+	std::mutex m_mutex;
+	std::mutex m_exception_mutex;
+
+	std::vector<QueuedPlayerData*> queue;
+};
