diff --git a/src/database/database-postgresql.cpp b/src/database/database-postgresql.cpp
index 3469f4242..5b733cba4 100644
--- a/src/database/database-postgresql.cpp
+++ b/src/database/database-postgresql.cpp
@@ -673,7 +673,7 @@ void AuthDatabasePostgreSQL::initStatements()
 {
 	prepareStatement("auth_read", "SELECT id, name, password, last_login FROM auth WHERE name = $1");
 	prepareStatement("auth_write", "UPDATE auth SET name = $1, password = $2, last_login = $3 WHERE id = $4");
-	prepareStatement("auth_create", "INSERT INTO auth (name, password, last_login) VALUES ($1, $2, $3) RETURNING id");
+	prepareStatement("auth_create", "INSERT INTO auth (name, password, last_login) VALUES ($1, $2, $3) ON CONFLICT(name) DO UPDATE SET name = excluded.name RETURNING id");
 	prepareStatement("auth_delete", "DELETE FROM auth WHERE name = $1");
 
 	prepareStatement("auth_list_names", "SELECT name FROM auth ORDER BY name DESC");
