diff --git a/src/constants.h b/src/constants.h
index 0a8b22e1..d6d297d4 100644
--- a/src/constants.h
+++ b/src/constants.h
@@ -52,9 +52,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 */
 
 // This many blocks are sent when player is building
-#define LIMITED_MAX_SIMULTANEOUS_BLOCK_SENDS 0
+#define LIMITED_MAX_SIMULTANEOUS_BLOCK_SENDS 10
 // Override for the previous one when distance of block is very low
-#define BLOCK_SEND_DISABLE_LIMITS_MAX_D 1
+#define BLOCK_SEND_DISABLE_LIMITS_MAX_D 2
 
 /*
     Map-related things
